operator/v1/types_csi_cluster_driver.go

```diff
@@ -61,6 +61,13 @@ const (
 // ClusterCSIDriverSpec is the desired behavior of CSI driver operator
 type ClusterCSIDriverSpec struct {
 	OperatorSpec `json:",inline"`
+	// DisableStorageClass determines if creation of storageclass for this driver
+	// should be skipped. By default Openshift creates storageclass for every CSI
+	// driver it installs via CSI operator. Setting this value to true will skip
+	// automatic creation of storageclass for driver being installed.
+	// Defaults to false.
+	// +kubebuilder:default=false
+	DisableStorageClass bool `json:"disableStorageClass,omitempty"`
```

Comment:
Bools are disallowed by convention. A lot of ideas start as a bool but in the future need additional states (for example "I don't care, you decide") which cannot be represented by a bool. This then leads to cases where users create multiple fields that depend on one another which is bad practice.

In general, rather than having a bool, it is better to have an enum that can be extended later.

For example, is there any configuration for the storage class that we may want to support at a later date? You may want something that looks like

```
storageClass:
  state: Managed | Unmanaged
  managed:
    <storageClass configuration for managed storage classes>
```

Reply:
I am not sure I like the idea (and there are plenty of examples of existing bools in API). If we must use an enum, something that sounds more suitable to me is - `storageClassState: managed|unmanaged|removed`. 

Having a nested `state` field within `storageClass` field is something I do no agree with, because then `storageClass` becomes an object which has a single property. I don't see us ever needing another field in `storageClass` type and if we are not going to need it - why define `storageClass` as an object? Seems unnecessary.

Reply:
> (and there are plenty of examples of existing bools in API)

While there are plenty of previous bools, we learn from our mistakes and we must not make those mistakes again. Note also, a lot of APIs didn't go through a thorough API review before they shipped early in OpenShift 4 so the API review team didn't get a chance to make recommendations not to use those. We are in the process of documenting and enforcing our conventions to make our future APIs better. New additions of bools are explicitly disallowed.

> `storageClassState: managed|unmanaged|removed`

This is also a valid suggestion, IF you are certain that you don't want that future expansion. I'm happy to allow that enum, though, if you ever come back and decide you do want to have any storage class configuration within this object, you will then have issues with the API design and trying to make something sensible that makes it obvious that you want to have the configuration valid in only certain scenarios, are you familiar with the concept of a discriminated union? 

This decision I'll leave up to you as the SME on this particular API. My example was just a suggestion

Note, as this enum is new, you must allow the value to be empty, you will need to install the updated API into an existing cluster and make sure that it doesn't break on upgrades before this PR merges.

Reply:
> This is also a valid suggestion,

ack - I do not think we have a valid use case of adding additional storageclass configuration parameters in `ClusterCSIDriver` objects. Basically this switch is pretty simple - if a user doesn't like storageclass OCP ships with, they can disable it and create their own. Exposing further knobs for storageclass configuration seems redundant (also because storageclasses aren't really all that complicated to configure and create).

> you will then have issues with the API design and trying to make something sensible that makes it obvious that you want to have the configuration valid in only certain scenarios, are you familiar with the concept of a discriminated union?

Yes - an earlier version of API I implemented for `ClusterCSIDriver` used driver names as union discriminators - https://github.com/openshift/api/commit/525a7c608de6175b7d8f840fd5e6757827f43797#diff-1890d43482146b467ff1905bfe70773c54f1bb865f348b2b93fd8f914a91bf09R67 . In a foreseeable future, I am not really worried about storageclass types which are going to need union discriminators for various cloudproviders. We should rather let them create their own SCs if they need and that is what this proposal does.

cc @jsafrane @dobsonj

Reply:
I'm fine with something like `storageClassState: Managed|Unmanaged|Removed` suggested above. I don't think storage class will need more configuration.